stdlib/json.gr

```diff
@@ -0,0 +1,2308 @@
+  let buf = Buffer.make(16)
+
+  let writer = makePrettyPrintJSONWriter(format, buf)
+
+  let error = writer.emit(json)
+
+  match (error) {
+    None => Ok(Buffer.toString(buf)),
+    Some(e) => Err(e),
+  }
+}
+
+/**
+ * Prints the JSON object into a string with compact formatting optimized for
+ * small size. Recommended for all uses where the intended consumer is a
+ * machine program. For example in REST APIs.
+ * 
+ * Using this function can be preferred over `toString` with compact formatting
+ * settings since it can be slightly faster.
+ * 
+ * @param json: The JSON object to print
+ * @returns A `Result` object with either the string containing the printed JSON or an error if the input object cannot be represented in the JSON format.
+ *
+ * @example print(Result.unwrap(toStringCompact(JSONObject([("currency", JSONString("€")), ("price", JSONNumber(99.9))]))))
+ * 
+ * Example output:
+ * ```json
+ * {"currency":"€","price":99.9}
+ * ```
+ */
+export let toStringCompact = (json: JSON) => {
+  let buf = Buffer.make(16)
+
+  let writer = makeCompactJSONWriter(buf)
+
+  let error = writer.emit(json)
+
+  match (error) {
+    None => Ok(Buffer.toString(buf)),
+    Some(e) => Err(e),
+  }
+}
+
+/**
+ * Prints the JSON object into a string with default pretty formatting
+ * settings. Recommended for uses where the intended consumer is a person or
+ * the text should be easily inspectable. For example configuration files or
+ * document file formats.
+ * 
+ * @param json: The JSON object to print
+ * @returns A `Result` object with either the string containing the printed JSON or an error if the input object cannot be represented in the JSON format.
+ * 
+ * Example output:
+ * ```json
+ * {
+ *   "currency": "€",
+ *   "price": 99.9
+ * }
+ * ```
+ *
+ * @example print(Result.unwrap(toStringPretty(JSONObject([("currency", JSONString("€")), ("price", JSONNumber(99.9))]))))
+ */
+export let toStringPretty = (json: JSON) =>
+  toString(json, defaultPrettyFormat())
+
+/**
+ * Represents errors for JSON parsing along with a human readable text message.
+ */
+export enum JSONParseError {
+  UnexpectedEndOfInput(String),
+  UnexpectedToken(String),
+  InvalidUTF16SurrogatePair(String),
+}
+
+/**
+ * Internal data structure used during parsing.
+ */
+record JSONParserImplHelper {
+  string: String,
+  bufferParse: Buffer.Buffer,
+  mut currentCodePoint: Number,
+  mut pos: Number,
+  mut bytePos: Number,
+}
+
+let isInterTokenWhiteSpace = (codePoint: Number) => {
+  match (codePoint) {
+    0x0009 => true, // tab
+    0x000A => true, // line feed
+    0x000D => true, // carriage return
+    0x0020 => true, // space
+    _ => false,
+  }
+}
+
+let _END_OF_INPUT = -1
+
+exception MalformedUnicode
+
+// This function has been copied from the String module as a temporary
+// solution. This will likely be replaced by a more robust solution for
+// iterating over strings in the near future.
+@disableGC
+let getCodePoint = (ptr: WasmI32) => {
```

Comment:
Once #1074 and associated PRs are merged, the `@disableGC` functions in this file should probably be migrated to `@unsafe`. @ospencer can you confirm that `@unsafe` functions shouldn't incur any excessive performance hit over their `@disableGC` counterparts, when all of the code is doing WasmXX-munging? `@unsafe` has the benefit of being much more maintainable.

Reply:
Done for all except `getCodePoint`, since it's a copy of the function from the String module, which still has not been migrated to `@unsafe`. I'd like to remove this duplicated function in the future, but first a replacement functionality needs to be implemented in stdlib (still TBD).